include/ensmallen_bits/iqn/iqn_impl.hpp

```diff
@@ -89,36 +113,37 @@ double IQN::Optimize(DecomposableFunctionType& function, arma::mat& iterate)
       const size_t effectiveBatchSize = std::min(batchSize, numFunctions -
           it * batchSize);
 
-      if (arma::norm(iterateVec - t.slice(it)) > 0)
+      if (arma::norm(iterate - t[it]) > 0)
       {
         function.Gradient(iterate, it * batchSize, gradient,
             effectiveBatchSize);
         gradient /= effectiveBatchSize;
 
-        const arma::mat s = iterateVec - t.slice(it);
-        const arma::mat yy = arma::vectorise(gradient - y.slice(it));
+        const BaseMatType s = arma::vectorise(iterate - t[it]);
+        const BaseGradType yy = arma::vectorise(gradient - y[it]);
 
-        const arma::mat stochasticHessian = Q.slice(it) + yy * yy.t() /
-            arma::as_scalar(yy.t() * s) - Q.slice(it) * s * s.t() *
-            Q.slice(it) / arma::as_scalar(s.t() * Q.slice(it) * s);
+        const BaseGradType stochasticHessian = Q[it] + yy * yy.t() /
+            arma::as_scalar(yy.t() * s) - Q[it] * s * s.t() *
+            Q[it] / arma::as_scalar(s.t() * Q[it] * s);
 
         // Update aggregate Hessian approximation.
-        B += (1.0 / numBatches) * (stochasticHessian - Q.slice(it));
+        B += (1.0 / numBatches) * (stochasticHessian - Q[it]);
 
         // Update aggregate Hessian-variable product.
-        u += (1.0 / numBatches) * (stochasticHessian * iterateVec -
-            Q.slice(it) * t.slice(it));
+        u += arma::reshape((1.0 / numBatches) * (stochasticHessian *
```

Comment:
Without thorough analysis, I don't know whether this call to reshape() recreates the given matrix with a different total number of elements (either less or more elements).  However, bear in mind that in the current implementation of SpMat::reshape(), the total number of elements can't be changed.

Reply:
Agreed---in this case, it is known that the number of elements will remain unchanged.